Issue #409 Refactoring start.go

hferentschik opened the pull request:
- Introducing ClusterUpConfig struct to separate Viper flags processing from 'cluster up' call
- Moving ClusterUp and PostClusterUp into own package under pks, since they are now independent of the command
- Updated tests

hferentschik commented:
Need to work out some integration issues ...

praveenkumar commented:
@hferentschik you need to rebase and solve those conflicts :)

hferentschik commented:
> you need to rebase and solve those conflicts :)

I did another rebase

praveenkumar commented on cmd/minishift/cmd/start_test.go:
```diff
@@ -116,57 +128,62 @@ func (r *RecordingRunner) Output(command string, args ...string) ([]byte, error)
-var testIp = "192.168.99.42"
+var (
+	testConfig = &clusterup.ClusterUpConfig{
+		OpenShiftVersion: "v1.5.1",
```
Finally version update for unit tests 👍

praveenkumar commented on pkg/minishift/oc/oc.go:
```diff
@@ -34,9 +37,9 @@ const (
-	runner         util.Runner
+	OcPath         string
+	KubeConfigPath string
+	Runner         util.Runner
```
👍

praveenkumar commented on pkg/testing/cli/harness.go:
```diff
@@ -76,6 +76,14 @@ func CreateExitHandlerFunc(t *testing.T, tee *Tee, expectedExitCode int, expecte
 	return exitHandler
 }
 
+func PreventExitHandlerFunc(t *testing.T) func(int) bool {
```
Does this mean it will return a function which have return type as `bool` and take a `int` parameter? Just want to understand.

hferentschik replied:
So what's happening in some tests is, that if they are setup wrong and the code actually exits (using an atexit method for example), the program/test execution will stop directly. This means that no assertions will be run marking the test effectively as passed. Even worse, subsequent tests won't run. We had this in other test as well. 

Using this handler an os.Exit will be converted to a panic. This means the test will actually fail (with a panic, but at least it does not get unnoticed). 

I will try to find better name and add some godocs. Hope it becomes clearer then

budhrg commented on cmd/minishift/cmd/addon/addon_apply.go:
```diff
@@ -75,15 +77,22 @@ func runApplyAddon(cmd *cobra.Command, args []string) {
-	kubeConfigPath := constants.KubeConfigPath
 	sshCommander := provision.GenericSSHCommander{Driver: host.Driver}
+	ocPath := minishiftConfig.InstanceConfig.OcPath
+	kubeConfigPath := constants.KubeConfigPath
```
Unless, `ocPath` or `kubeConfigPath` is used multiple times in later part of this file, we can directly pass these values.

hferentschik replied:
hmm, to humour you. Usually, I prefer shorter lines.

budhrg commented on cmd/minishift/cmd/addon/util.go:
```diff
@@ -20,21 +20,14 @@ import (
-)
-
-// getAddOnManager returns the addon manager
+// GetAddOnManager returns the addon manager
```
Same change I did in my PR, looks like a conflict when I do rebase :wink:

hferentschik replied:
:-)

budhrg commented on cmd/minishift/cmd/start.go:
```diff
@@ -115,36 +108,16 @@ func runStart(cmd *cobra.Command, args []string) {
-
-	fmt.Printf("Starting local OpenShift cluster using '%s' hypervisor...\n", machineConfig.VMDriver)
-
+	// to determine whether we need to run post cluster up actions,
```
`to determine` to `Determine`. We have started comments with Caps (at least I am following it).

budhrg commented on cmd/minishift/cmd/start.go:
```diff
@@ -115,36 +108,16 @@ func runStart(cmd *cobra.Command, args []string) {
+	hostVm, ip := startHost(libMachineClient)
+
+	if proxyConfig.IsEnabled() {
+		// once we know the IP, we need to make sure it is not proxied in a proxy environment
```
Same here, `once we know` to `Once we know`.

budhrg commented on cmd/minishift/cmd/start.go:
```diff
@@ -154,42 +127,50 @@ func runStart(cmd *cobra.Command, args []string) {
+	clusterUpParams := determineClusterUpParameters(clusterUpConfig)
+	err = clusterup.ClusterUp(clusterUpConfig, clusterUpParams, &util.RealRunner{})
+	if err != nil {
+		atexit.ExitWithMessage(1, fmt.Sprintf("Error during'cluster up' execution: %v", err))
```
Space missing after `Error during` word.

budhrg commented on cmd/minishift/cmd/start.go:
```diff
@@ -319,29 +243,15 @@ func automountHostfolders(driver drivers.Driver) {
-
-	provision.SetDetector(&provisioner.MinishiftProvisionerDetector{Delegate: provision.StandardDetector{}})
-}
+// initStartFlags creates the CLI flags which needs to be passed on to 'lib-machine'
```
`lib-machine` to `libmachine`. I think `libmachine` is more familiar than `lib-machine`.

hferentschik replied:
fair enough

budhrg commented on cmd/minishift/cmd/start.go:
```diff
@@ -371,44 +285,31 @@ func initClusterUpFlags() {
+	return clusterUpFlagSet
 }
 
 // initProxyFlags create the CLI flags which needs to be passed for proxy
```
Wrong comment.

hferentschik replied:
good catch

budhrg commented on cmd/minishift/cmd/start.go:
```diff
@@ -371,44 +285,31 @@ func initClusterUpFlags() {
 // initProxyFlags create the CLI flags which needs to be passed for proxy
-func initSubscriptionManagerFlags() {
+func initSubscriptionManagerFlags() *flag.FlagSet {
+	var subscriptionManagerFlagSet = flag.NewFlagSet(commandName, flag.ContinueOnError)
```
Why not `subscriptionManagerFlagSet := flag.NewFlagSet(commandName, flag.ContinueOnError)`?

hferentschik replied:
same same really. I guess I copied the 'var subscriptionManagerFlagSet' part originally from the top of the file. 

I guess we usually use the :=, so I am going to change it

budhrg commented on cmd/minishift/cmd/start_test.go:
```diff
@@ -116,57 +128,62 @@ func (r *RecordingRunner) Output(command string, args ...string) ([]byte, error)
+	testConfig = &clusterup.ClusterUpConfig{
+		OpenShiftVersion: "v1.5.1",
+		Ip:               "192.168.99.42",
+		OcPath:           "/home/john/.minishift/cache/oc/1.5.0/oc",
```
`OcPath` should have `1.5.1` as version.

hferentschik replied:
that's really irrelevant here. I could have used /foo/bar here. In fact I was doing so initially, partly to make clear that this is not a real oc path. I am changing it to 1.5.1, but also considering using a pure dummy string.

budhrg replied:
@hferentschik , if that is the case, let's put dummy string here then. This path make me feel like we might be coming some version here :)

budhrg commented on pkg/minishift/clusterup/clusterup.go:
```diff
@@ -0,0 +1,227 @@
+)
+
+const (
+	ip_key             = "ip"
```
Let change `ip_key` to `ipKey` and `routing_suffix_key` to `routingSuffixKey`.

budhrg commented on pkg/minishift/clusterup/clusterup.go:
```diff
@@ -0,0 +1,227 @@
+// PostClusterUp runs the Minishift specific provisioning after 'cluster up' has run
+func PostClusterUp(clusterUpConfig *ClusterUpConfig, sshCommander provision.SSHCommander, addOnManager *manager.AddOnManager) error {
+	err := kubeconfig.CacheSystemAdminEntries(clusterUpConfig.KubeConfigPath, getConfigClusterName(clusterUpConfig.Ip, clusterUpConfig.Port))
+	if err != nil {
```
```
if err != nil {
   return err
}
```
is the repeating pattern in this method. 

Seems to be some kind of code smelling here. Let do some refactoring here.

hferentschik replied:
Is it? That's just usual Go error handling to me. I am intrigued to know how you want to refactor this? I making various which potentially can return an error. I need to check whether an error occurred and act accordingly. In this case I just want to pass it on.

budhrg replied:
I am thinking on this line.
```
func checkAndReturnError(err error) {
       if err != nil {
            return err
       }
}

// Call above instead of three lines
```
Choose the name accordingly :smile:

budhrg replied:
Fair enough.
At least, if possible let's use this pattern : 
```
if err := ocRunner.AddSudoerRoleForUser(clusterUpConfig.User); err != nil {
	return err
}
```

hferentschik replied:
not for me. If you prefer this syntax go for it, but I don't like it at all.

budhrg commented on pkg/minishift/clusterup/clusterup.go:
```diff
@@ -0,0 +1,227 @@
+	return nil
+}
+
+//EnsureHostDirectoriesExist Ensures the specified directories exist on the virtual machine and creates them if not
```
Space missing in between `//` and `EnsureHostDirectoriesExist`

Also, comment should be
```
// EnsureHostDirectoriesExist ensures that the specified directories exist on the VM and creates them if does not exist
```

budhrg commented on pkg/minishift/clusterup/clusterup.go:
```diff
@@ -0,0 +1,227 @@
+}
+
+// DetermineOpenShiftVersion returns the OpenShift/oc version to use. If the requested version is
+// < the base line version we use baseline oc to provision the requested OpenShift version. If the requested
```
Comma missing after `< the base line version`. Took me few  seconds to process this line :smile:  
Also, I feel if we write the comments one comment per line, then it will increase readability.

```
DetermineOpenShiftVersion returns the OpenShift/oc version to use. 
If the requested version is < the base line version we use baseline oc to provision the requested OpenShift version. 
If the requested OpenShift version is >= the baseline, we align the oc version with the requested OpenShift version.
```

budhrg commented on pkg/minishift/clusterup/clusterup.go:
```diff
@@ -0,0 +1,227 @@
+	return requestedVersion
+}
+
+func GetExecutionContext(ip string, routingSuffix string, ocRunner *oc.OcRunner, sshCommander provision.SSHCommander) (*command.ExecutionContext, error) {
```
Comment here will help more to understand this method well.

budhrg commented on pkg/minishift/clusterup/clusterup.go:
```diff
@@ -0,0 +1,230 @@
+// DetermineOpenShiftVersion returns the OpenShift/oc version to use.
+// If the requested version is < the base line version we use baseline oc to provision the requested OpenShift version.
+// If the requested OpenShift version is >= the baseline, we align the oc version with the requested OpenShift version.
+// It also handles the case the user just specifies a version without the required 'v' prefix
```
:+1:

hferentschik replied:
reads indeed better :-)

budhrg commented on pkg/minishift/clusterup/clusterup.go:
```diff
@@ -0,0 +1,230 @@
+	return requestedVersion
+}
+
+// GetExecutionContext createds an ExecutionContext for executing add-on.
```
Typo here @hferentschik  :smile:

hferentschik replied:
got it, thanks

budhrg commented:
@hferentschik , I am done with PR review. Please fix the typo and this change https://github.com/minishift/minishift/pull/962#discussion_r119270241 .

Rest LGTM :+1:

LalatenduMohanty commented on cmd/minishift/cmd/addon/addon_apply.go:
```diff
@@ -75,15 +77,20 @@ func runApplyAddon(cmd *cobra.Command, args []string) {
-		addonContext := GetExecutionContext(ip, routingSuffix, ocPath, kubeConfigPath, sshCommander)
+		addonContext, err := clusterup.GetExecutionContext(ip, routingSuffix, ocRunner, sshCommander)
+		if err != nil {
+			atexit.ExitWithMessage(1, fmt.Sprint("Error executing addon commands: ", err))
```
I think you missed the `%s`.

praveenkumar replied:
@LalatenduMohanty No %s required when you use `Sprintf`

hferentschik replied:
Right, not required here. Different method as @praveenkumar is saying

gbraad replied:
@LalatenduMohanty you can actually write `"Hello", "World", "!"`
https://play.golang.org/p/3Mqog0waoH

LalatenduMohanty replied:
Got it.  I thought it is `Sprintf`

LalatenduMohanty commented:
The code looks good. I did test the `minishift start` on KVM and it worked fine.

gbraad commented on cmd/minishift/cmd/addon/addon_apply.go:
```diff
@@ -75,15 +77,20 @@ func runApplyAddon(cmd *cobra.Command, args []string) {
 	routingSuffix := viper.GetString(routingSuffix)
-	kubeConfigPath := constants.KubeConfigPath
 	sshCommander := provision.GenericSSHCommander{Driver: host.Driver}
+	ocRunner, err := oc.NewOcRunner(minishiftConfig.InstanceConfig.OcPath, constants.KubeConfigPath)
```
I wonder why this is not used in the tests. At the moment we use: 	`"github.com/minishift/minishift/test/integration/util"`

hferentschik replied:
It should be now. I think historically the "stub" OcRunner in the integration test was there first. The fully fleshed out runner as part of the production code is not around for so long. So yes, it could/should be used in the integration tests.

hferentschik commented:
Pushed another update.

gbraad commented on cmd/minishift/cmd/start.go:
```diff
@@ -83,68 +76,49 @@ var (
 	libMachineClient := libmachine.NewClient(constants.Minipath, constants.MakeMiniPath("certs"))
 	defer libMachineClient.Close()
 
+	ensureNotRunning(libMachineClient, constants.MachineName)
```
+1 (does it do error reporting?)

hferentschik replied:
It is part of the pre-flight checks we recently discussed. I noticed that if I had a machine running and did a 'minishift start', it would still do cluster up stuff. Effectively to no prevail. This is a check which will check whether the VM is running and if so, it will print a message of the form "Minishift VM already running".

gbraad commented on cmd/minishift/cmd/start.go:
```diff
@@ -154,42 +128,50 @@ func runStart(cmd *cobra.Command, args []string) {
-	if err != nil {
-		fmt.Println("Error creating required host directories: ", err)
-		atexit.Exit(1)
+	hostDirs := []string{
```
I prefer to extract to function.

hferentschik replied:
Is clusterup.EnsureHostDirectoriesExist not the function you are after? The thing is, I need to provide this function with the set of directories to check/create. you want me to put 

	hostDirs := []string{
		viper.GetString(startFlags.HostConfigDir.Name),
		viper.GetString(startFlags.HostDataDir.Name),
		viper.GetString(startFlags.HostVolumeDir.Name),
		viper.GetString(startFlags.HostPvDir.Name),
	}

into a separate function?

hferentschik replied:
I put it into a function. I guess the benefit is to a certain degree the gain in clarity the function name provides.

gbraad commented on cmd/minishift/cmd/start.go:
```diff
@@ -154,42 +128,50 @@ func runStart(cmd *cobra.Command, args []string) {
 	}
-
-	ip, err := host.Driver.GetIP()
+	err = clusterup.EnsureHostDirectoriesExist(hostVm, hostDirs)
```
... which can be used as the input for this function

hferentschik replied:
got you. Done.

gbraad commented on cmd/minishift/cmd/start.go:
```diff
@@ -154,42 +128,50 @@ func runStart(cmd *cobra.Command, args []string) {
 	if err != nil {
-		fmt.Println("Error determining host ip: ", err)
-		atexit.Exit(1)
+		atexit.ExitWithMessage(1, fmt.Sprintf("Error creating required host directories: %v", err))
```
and extract even further or combine as:
```
    if err := clusterup.EnsureHostDirectoriesExist(hostVm, getHostDirsFromViper()); err != nil {
        atexit.ExitWithMessage(1, fmt.Sprintf("Error creating required host directories: %v", err))
    }
```

gbraad commented on cmd/minishift/cmd/start.go:
```diff
@@ -154,42 +128,50 @@ func runStart(cmd *cobra.Command, args []string) {
-		proxyConfig.ApplyToEnvironment()
+	autoMountHostFolders(hostVm.Driver)
+
+	openShiftVersion := clusterup.DetermineOpenShiftVersion(viper.GetString(startFlags.OpenshiftVersion.Name))
```
Suggest extracting into a meaningful function...

hferentschik replied:
What do you mean? What do you want me to extract?

gbraad replied:
```
function determineOpenShiftVersion() {
    return clusterup.DetermineOpenShiftVersion(viper.GetString(startFlags.OpenshiftVersion.Name))
}
```

hferentschik replied:
I am lost here. We are are talking about a single line here, and you want me to make three lines and a function call out of this. AFAICT it also does not increase clarity here. 

> as it stands now, the GetString is likely used to override the value we seed. 

It does not override anything, it just gets it from Viper. The return value from clusterup.DetermineOpenShiftVersion does an override of sort, but I think that is well documented in the Godocs of clusterup.DetermineOpenShiftVersion.

gbraad commented on cmd/minishift/cmd/start.go:
```diff
@@ -154,42 +128,50 @@ func runStart(cmd *cobra.Command, args []string) {
+		User:             defaultUser,
+		Project:          defaultProject,
+		KubeConfigPath:   constants.KubeConfigPath,
+		OcPath:           cacheOc(openShiftVersion),
```
... and use here; as it is likely the only place where it is used.

hferentschik replied:
In case you are talking about openShiftVersion, it is used twice

gbraad replied:
I see yeha, maybe the `cacheOc` should not happen here. It is a function that does too much

hferentschik replied:
Sure, I can do that as well. In this case it becomes something like:

       ocPath := cacheOc(openShiftVersion)
	clusterUpConfig := &clusterup.ClusterUpConfig{
		OpenShiftVersion: openShiftVersion,
		MachineName:      constants.MachineName,
		Ip:               ip,
		Port:             constants.APIServerPort,
		RoutingSuffix:    viper.GetString(startFlags.RoutingSuffix.Name),
		HostPvDir:        viper.GetString(startFlags.HostPvDir.Name),
		User:             defaultUser,
		Project:          defaultProject,
		KubeConfigPath:   constants.KubeConfigPath,
		OcPath:           ocPath,
	}

TBH, I prefer this as well. However, I got several other comments where people thing it is "better" to rather inlines a function call into another function call, over having two separate lines.

> maybe the cacheOc should not happen here. It is a function that does too much

it needs to happen here or just before creating the struct. _ocPath_ downloads and caches the _oc_ binary based on the OpenShift version we are after. It returns the path to the cached binary. I am not sure why you think it is doing too much.

That said, I am fine to split this into two lines. Note though, that I still need the openShiftVersion twice.

gbraad commented on cmd/minishift/cmd/start.go:
```diff
@@ -154,42 +128,50 @@ func runStart(cmd *cobra.Command, args []string) {
-
-	clusterUp(&machineConfig, ip)
+	clusterUpParams := determineClusterUpParameters(clusterUpConfig)
+	err = clusterup.ClusterUp(clusterUpConfig, clusterUpParams, &util.RealRunner{})
```
Make the code more stand out as related
```
if err := function(); err != nil {
   ...
}
```

hferentschik replied:
I am not a big fan of this syntax for various reasons

gbraad replied:
but you do use it later on... it makes the code more cohesive.

hferentschik replied:
I usually never use this syntax. I actually think the majority of code is using the style I am using here.

gbraad commented on cmd/minishift/cmd/start.go:
```diff
@@ -154,42 +128,50 @@ func runStart(cmd *cobra.Command, args []string) {
+		atexit.ExitWithMessage(1, fmt.Sprintf("Error during 'cluster up' execution: %v", err))
+	}
 
 	if !isRestart {
```
create a function for this...

hferentschik replied:
for what? the content in the if statement?

gbraad replied:
Yes . It seems to be something like `performPostClusterUpConfiguration() {}`

hferentschik replied:
+1

hferentschik replied:
done

gbraad commented on cmd/minishift/cmd/start.go:
```diff
@@ -215,101 +197,44 @@ func handleProxies() *util.ProxyConfig {
 }
 
-func automountHostfolders(driver drivers.Driver) {
+func autoMountHostFolders(driver drivers.Driver) {
```
so you did want to change it ;-). had this same question when working on the task.

hferentschik replied:
:-)

gbraad commented on cmd/minishift/cmd/start.go:
```diff
@@ -439,35 +353,35 @@ func setDefaultRoutingPrefix(ip string) {
-				constants.MinOpenshiftSuportedVersion)
-			atexit.Exit(1)
-		}
+func ensureNotRunning(client *libmachine.Client, machineName string) {
```
so the function does exit... I consider this kind of a side-effect. however the name `Ensure` kind of implies this...

hferentschik replied:
Not quite sure what you are saying.

gbraad replied:
@hferentschik related to an earlier comment: https://github.com/minishift/minishift/pull/962#discussion_r119299583

hferentschik commented:
@gbraad I pushed another update addressing most of your comments. I think I commented also on all of your comments, but if I missed something let me know.

LalatenduMohanty commented:
@gbraad Going to merge this as your comments are addressed.
